sympy/integrals/laplace.py

```diff
@@ -1089,6 +1096,208 @@ def simp_exp(arg):
     return _simplify(f.subs(t, t_), simplify), cond
 
 
+def _complete_the_square_in_denom(f, s):
+    from sympy.simplify.radsimp import fraction
+    [n, d] = fraction(f)
+    if d.is_polynomial(s):
+        cf = d.as_poly(s).all_coeffs()
+        if len(cf)==3:
+            a, b, c = cf
+            d = a*((s+b/(2*a))**2+c/a-(b/(2*a))**2)
+    return n/d
+
+
+@cacheit
+def _inverse_laplace_build_rules():
+    """
+    This is an internal helper function that returns the table of inverse
+    Laplace transform rules in terms of the time variable `t` and the
+    frequency variable `s`.  It is used by `_inverse_laplace_apply_rules`.
+    """
+    s = Dummy('s')
+    t = Dummy('t')
+    a = Wild('a', exclude=[s])
+    b = Wild('b', exclude=[s])
+    c = Wild('c', exclude=[s])
+    d = Wild('d', exclude=[s])
+
+    debug('_inverse_laplace_build_rules is building rules')
+    def _frac(f, s):
+        try:
+            return f.factor(s)
+        except PolynomialError:
+            return f
+    same = lambda f: f
+    frac = lambda f: _frac(f, s)
+    ctsd = lambda f: _complete_the_square_in_denom(f, s)
+    # This list is sorted according to the prep function needed.
+    _ILT_rules = [
+        (a/s, a, S.true, same, 1),
+        (b/(s+a), b*exp(-a*t), S.true, same, 1),
+        (a/s**2, a*t, S.true, same, 1),
+        (b/(s**2+a**2), b*sin(a*t)/a, S.true, same, 1),
+        (b/(s**2-a**2), b*sinh(a*t)/a, S.true, same, 1),
+        (b*s/(s**2+a**2), b*cos(a*t), S.true, same, 1),
+        (b*s/(s**2-a**2), b*cosh(a*t), S.true, same, 1),
+        (b*(s+a)**(-c), t**(c-1)*exp(-a*t)/gamma(c), c>0, same, 1),
+        #
+        (b/(s*(s+a)), b*(1-exp(-a*t))/a, S.true, frac, 1),
+        (b/(s+a)**2, b*t*exp(-a*t), S.true, frac, 1),
+        (b*s/(s+a)**2, b*(1-a*t)*exp(-a*t), S.true, frac, 1),
+        (c/((s+a)*(s+b)), c/(a-b)*(exp(-b*t)-exp(-a*t)), S.true, frac, 1),
+        (c*s/((s+a)*(s+b)), c/(a-b)*(a*exp(-a*t)-b*exp(-b*t)), S.true, frac, 1),
+        #
+        (c/(d*(s+b)**2+a**2),
+         c*exp(-b*t)*sin(a*t/sqrt(d))/(a*sqrt(d)), S.true, ctsd, 1),
+        (c/(d**2*(s+a)**2+b**2),
+         c/d*exp(-a*t)*(b*cos(b/d*t)/d-a*sin(b/d*t))/b,
+         S.true, ctsd, 1/s),
+    ]
+    return _ILT_rules, s, t
+
+
+def _inverse_laplace_apply_simple_rules(f, s, t):
+    """
+    Helper function for the class InverseLaplaceTransform.
+    """
+    if f==1:
+        debug('_inverse_laplace_apply_simple_rules match:')
+        debug('      f:    %s'%(1,))
+        debug('      rule: 1 o---o DiracDelta(%s)'%(t,))
+        return DiracDelta(t), S.true
+
+    _ILT_rules, s_, t_ = _inverse_laplace_build_rules()
+    _prep = ''
+
+    for s_dom, t_dom, check, prep, fac in _ILT_rules:
+        if not _prep is (prep, fac):
+            _F = prep(f.subs({s: s_})*fac)
```

Comment:
I think that ideally, the substitution should move more outside. It should be sufficient to be run only once in `doit` and substituted back maybe before passing to `integrate` fallback to users. It complicates things a bit, but should be doable.

We should make the table to build two staged, first the dummy symbols can be built, and then the table can be built with respect to the the cached symbols only when the table is looked up in first time, 
Or I'm also thinking of other ways, like making every entries of the table a function than value `() -> ...` that should evaluate to give the actual expression, or so on.

Reply:
> I think that ideally, the substitution should move more outside.

I would like to do that late in my development list. While I still have to add rules, and have to decide whether they are simple or programmed, I am quite happy if the code is as uncomplicated as possible. I have seen that caching saves a lot of time, but this next step will not save a lot. OK?

Reply:
Just for your reference, in LT, less than 10% of the known rules are implemented ...

Reply:
The substitution and unsubstitution should be done in precisely one place so that it does not need to be kept track of in many places. Otherwise it will be forgotten somewhere.

Reply:
> The substitution and unsubstitution should be done in precisely one place so that it does not need to be kept track of in many places. Otherwise it will be forgotten somewhere.

I totally agree. I understood @sylee957 as both could be moved outside. I don't yet see how this would be beneficial, though.

Reply:
t could be done at the start and end of `_inverse_laplace_transform`. Otherwise if it is only needed for rule matching then it it could be done here but it shouldn't be done repeatedly in a loop.

Reply:
What does `if not _prep is (prep, fac)` do? Are you sure it works as intended? Should it not at least be `if not _prep == (prep, fac)`?

Reply:
It works as intended, but is wrong, `==` is what I want.